src/analytics/lambdas/refresh-materialized-views-workflow/check-start-sp-refresh.ts

```diff
@@ -0,0 +1,171 @@
+    if (event.detail.completeRefreshDate) {
+      // force refresh has been ended
+      return {
+        detail: {
+          nextStep: END_STEP,
+        },
+      };
+    } else if (event.originalInput.startRefreshViewOrSp) {
+      const index = spList.findIndex((spInfo) => spInfo.name === event.originalInput.startRefreshViewOrSp);
+      if (index !== -1) {
+        return {
+          detail: {
+            startRefreshViewOrSp: event.originalInput.startRefreshViewOrSp,
+            refreshDate: refreshDateString,
+            appId: timezoneWithAppId.appId,
+            timezone: timezoneWithAppId.timezone,
+            forceRefresh: event.originalInput.forceRefresh,
+            nextStep: REFRESH_SP_STEP,
+          },
+        };
+      } else {
+        // no sp need to be force refreshed
+        return {
+          detail: {
+            nextStep: END_STEP,
+          },
+        };
+      }
+    } else {
+      throw new Error('forceRefresh is true, but no completeRefreshView or startRefreshView found');
+    }
+  } else {
+    if (event.detail.completeRefreshDate) {
+      // check whether there need to refresh earlier date
+      const date = new Date(event.detail.completeRefreshDate);
+      if (date > refreshEarliestDate) {
+        date.setDate(date.getDate() - 1);
+        return {
+          detail: {
+            refreshDate: date.toISOString().split('T')[0],
+            appId: timezoneWithAppId.appId,
+            timezone: timezoneWithAppId.timezone,
+            nextStep: REFRESH_SP_STEP,
+          },
+        };
+      } else {
+        return {
+          detail: {
+            nextStep: END_STEP,
+          },
+        };
+      }
+    } else {
+      // first time to refresh sp
+      // check max refresh date
+      const propertyListSqlStatements: string[] = [];
+      propertyListSqlStatements.push(`SELECT MAX(refresh_date) FROM ${timezoneWithAppId.appId}.refresh_mv_sp_status where triggerred_by = 'WORK_FLOW';`);
+      const propertyListQueryId = await executeStatementsWithWait(
+        redshiftDataApiClient, propertyListSqlStatements, redshiftProps.serverlessRedshiftProps, redshiftProps.provisionedRedshiftProps);
+      const response = await getStatementResult(redshiftDataApiClient, propertyListQueryId!);
+      let lastRefreshDateStr;
+      if (response.Records) {
+        lastRefreshDateStr = response.Records[0][0]?.stringValue;
+      }
+      if (!lastRefreshDateStr || new Date(lastRefreshDateStr) < new Date(refreshDateString)) {
+        return {
+          detail: {
+            refreshDate: refreshDateString,
+            appId: timezoneWithAppId.appId,
+            timezone: timezoneWithAppId.timezone,
+            nextStep: REFRESH_SP_STEP,
+          },
+        };
+      } else {
+        return {
+          detail: {
+            nextStep: END_STEP,
+          },
+        };
+      }
+    }
+  }
+};
+
+function getRefreshEarliestDate(dataFreshnessInHour: string, refreshDate: string): Date {
```

Comment:
only support hours interval?

Reply:
This data is the same with dataFreshnessInHourParam in data-pipeline stack. 
Control plane can choose hour or day, control plane will convert to hour and send to the stack.